File: ios/calorietracker/Insights/TrainingLoad.swift
import Foundation

// Training load: overlapping workouts merged into sessions, minutes × intensity per local day, compared with the
// 28-day mean. Port of the "Training load" section of `scripts/insights_reference.py`.

nonisolated enum TrainingLoadEngine {
    struct Session: Sendable {
        var startMs: Int64
        var endMs: Int64
        var intensity: Double
        var hasEffort: Bool
        var members: Int
        var day: String = ""
        var minutes: Double = 0
        var load: Double = 0
    }

    struct DayLoad: Sendable {
        var load: Double = 0
        var minutes: Double = 0
        var sessions: Int = 0
    }

    struct Raw: Sendable {
        var day: String
        var load: Double
        var minutes: Double
        var sessions: Int
        var mean28d: Double
        var ratio: Double?
        var category: String
    }

    /// Workouts sorted by (start, end); an overlapping workout extends the previous session and keeps the max
    /// intensity. Day = local day of the session start.
    static func sessions(_ workouts: [InsightsWorkout], timeZone: TimeZone, config: InsightsConfig) -> [Session] {
        let t = config.trainingLoad
        let ws = workouts.filter { $0.endMs > $0.startMs }.enumerated().sorted { a, b in
            if a.element.startMs != b.element.startMs { return a.element.startMs < b.element.startMs }
            if a.element.endMs != b.element.endMs { return a.element.endMs < b.element.endMs }
            return a.offset < b.offset
        }.map(\.element)
        var out: [Session] = []
        for w in ws {
            let hasEffort = w.effort != nil
            let inten = w.effort.map { InsightsMath.clamp($0 / t.effortDivisor, t.intensityMin, t.intensityMax) }
                ?? t.defaultIntensity
            if let last = out.last, w.startMs < last.endMs {
                var s = last
                s.endMs = max(s.endMs, w.endMs)
                s.intensity = max(s.intensity, inten)
                s.hasEffort = s.hasEffort || hasEffort
                s.members += 1
                out[out.count - 1] = s
            } else {
                out.append(Session(startMs: w.startMs, endMs: w.endMs, intensity: inten, hasEffort: hasEffort, members: 1))
            }
        }
        for i in out.indices {
            out[i].day = InsightsDay.localDay(ms: out[i].startMs, timeZone)
            out[i].minutes = Double(out[i].endMs - out[i].startMs) / 60000.0
            out[i].load = out[i].minutes * out[i].intensity
        }
        return out
    }

    static func sessions(_ workouts: [InsightsWorkout], timeZone: String, config: InsightsConfig) -> [Session] {
        sessions(workouts, timeZone: InsightsDay.timeZone(timeZone), config: config)
    }

    static func dailyLoads(_ workouts: [InsightsWorkout], timeZone: TimeZone, config: InsightsConfig) -> [String: DayLoad] {
        var days: [String: DayLoad] = [:]
        for s in sessions(workouts, timeZone: timeZone, config: config) {
            var d = days[s.day] ?? DayLoad()
            d.load += s.load
            d.minutes += s.minutes
            d.sessions += 1
            days[s.day] = d
        }
        return days
    }

    static func loadRaw(_ loads: [String: DayLoad], _ day: String, config: InsightsConfig) -> Raw {
        let t = config.trainingLoad
        let today = loads[day] ?? DayLoad()
        var total = 0.0
        for k in stride(from: t.meanWindowDays, to: 0, by: -1) {
            total += loads[InsightsDay.add(day, -k)]?.load ?? 0.0
        }
        let avg = total / Double(t.meanWindowDays)
        let ratio: Double? = avg > 0 ? today.load / avg : nil
        let cat: String
        if today.load == 0 {
            cat = "none"
        } else if ratio == nil || ratio! > t.highAbove {
            cat = "high"
        } else if ratio! < t.lightBelow {
            cat = "light"
        } else {
            cat = "moderate"
        }
        return Raw(day: day, load: today.load, minutes: today.minutes, sessions: today.sessions, mean28d: avg,
                   ratio: ratio, category: cat)
    }

    /// Load of `day` (sum of session minutes × intensity) vs the mean daily load of the 28 days before it.
    static func trainingLoad(_ workouts: [InsightsWorkout], day: String, timeZone: String, config: InsightsConfig) -> TrainingLoadResult {
        let r = loadRaw(dailyLoads(workouts, timeZone: InsightsDay.timeZone(timeZone), config: config), day, config: config)
        return TrainingLoadResult(
            day: r.day, load: InsightsMath.roundTo(r.load, 1), minutes: InsightsMath.roundTo(r.minutes, 1),
            sessions: r.sessions, mean28d: InsightsMath.roundTo(r.mean28d, 1), ratio: InsightsMath.roundTo(r.ratio, 2),
            category: r.category, label: config.trainingLoad.labels[r.category] ?? r.category)
    }
}

File: android/app/src/main/java/com/ayuvo/health/insights/TrainingLoad.kt
package com.ayuvo.health.insights

import com.ayuvo.health.insights.InsightsMath.clamp
import com.ayuvo.health.insights.InsightsMath.roundTo
import java.time.LocalDate
import java.time.ZoneId

/**
 * Training load (docs/insights.md, "Training load"): overlapping workouts merge into sessions, a
 * session's load is minutes × intensity, and a day's load is compared with the mean daily load of
 * the 28 days before it. Ported from `sessions`, `daily_loads`, `_load_raw` and `training_load`.
 */
object TrainingLoad {

    data class Session(
        val startMs: Long,
        var endMs: Long,
        var intensity: Double,
        var hasEffort: Boolean,
        var members: Int
    ) {
        lateinit var day: LocalDate
        var minutes: Double = 0.0
        var load: Double = 0.0
    }

    data class DayLoad(val load: Double, val minutes: Double, val sessions: Int)

    data class Raw(
        val day: LocalDate,
        val load: Double,
        val minutes: Double,
        val sessions: Int,
        val mean28d: Double,
        val ratio: Double?,
        val category: String
    )

    /** Sorted by start then end; a session spans the union of its members and keeps the highest intensity. */
    fun sessions(workouts: List<WorkoutInput>, zone: ZoneId, cfg: InsightsConfig): List<Session> {
        val t = cfg.trainingLoad
        val ws = workouts.filter { it.endMs > it.startMs }.sortedWith(compareBy({ it.startMs }, { it.endMs }))
        val out = ArrayList<Session>()
        for (w in ws) {
            val hasEffort = w.effort != null
            val intensity = if (w.effort != null) clamp(w.effort / t.effortDivisor, t.intensityMin, t.intensityMax) else t.defaultIntensity
            val last = out.lastOrNull()
            if (last != null && w.startMs < last.endMs) {
                last.endMs = maxOf(last.endMs, w.endMs)
                last.intensity = maxOf(last.intensity, intensity)
                last.hasEffort = last.hasEffort || hasEffort
                last.members += 1
            } else {
                out += Session(w.startMs, w.endMs, intensity, hasEffort, 1)
            }
        }
        for (s in out) {
            s.day = InsightsMath.localDayOf(s.startMs, zone)
            s.minutes = (s.endMs - s.startMs) / 60000.0
            s.load = s.minutes * s.intensity
        }
        return out
    }

    fun dailyLoads(workouts: List<WorkoutInput>, zone: ZoneId, cfg: InsightsConfig): Map<LocalDate, DayLoad> {
        val days = LinkedHashMap<LocalDate, DayLoad>()
        for (s in sessions(workouts, zone, cfg)) {
            val d = days[s.day] ?: DayLoad(0.0, 0.0, 0)
            days[s.day] = DayLoad(d.load + s.load, d.minutes + s.minutes, d.sessions + 1)
        }
        return days
    }

    fun raw(loads: Map<LocalDate, DayLoad>, day: LocalDate, cfg: InsightsConfig): Raw {
        val t = cfg.trainingLoad
        val today = loads[day] ?: DayLoad(0.0, 0.0, 0)
        var total = 0.0
        for (k in t.meanWindowDays downTo 1) total += loads[day.minusDays(k.toLong())]?.load ?: 0.0
        val avg = total / t.meanWindowDays
        val ratio = if (avg > 0) today.load / avg else null
        val category = when {
            today.load == 0.0 -> "none"
            ratio == null || ratio > t.highAbove -> "high"
            ratio < t.lightBelow -> "light"
            else -> "moderate"
        }
        return Raw(day, today.load, today.minutes, today.sessions, avg, ratio, category)
    }

    fun trainingLoad(workouts: List<WorkoutInput>, day: LocalDate, zone: ZoneId, cfg: InsightsConfig): TrainingLoadResult {
        val r = raw(dailyLoads(workouts, zone, cfg), day, cfg)
        return TrainingLoadResult(
            day = r.day, load = roundTo(r.load, 1), minutes = roundTo(r.minutes, 1), sessions = r.sessions,
            mean28d = roundTo(r.mean28d, 1), ratio = roundTo(r.ratio, 2), category = r.category,
            label = cfg.trainingLoad.labels.getValue(r.category)
        )
    }
}
